File: on-prem-connector/ad_client.py
"""
Active Directory Client (LDAP)
Handles user sync and org structure from AD
"""

import logging
from typing import Optional
from ldap3 import Server, Connection, ALL, NTLM, SUBTREE
from ldap3.core.exceptions import LDAPException
import base64

logger = logging.getLogger(__name__)


class ADClient:
    def __init__(self, config: dict):
        self.config = config['ad']
        self.conn: Optional[Connection] = None

    def connect(self) -> bool:
        """Establish connection to AD"""
        try:
            server = Server(
                self.config['server'],
                port=self.config['port'],
                use_ssl=self.config['use_ssl'],
                get_info=ALL
            )

            self.conn = Connection(
                server,
                user=self.config['bind_user'],
                password=self.config['bind_password'],
                authentication=NTLM,
                auto_bind=True
            )

            logger.info(f"Connected to AD: {self.config['server']}")
            return True

        except LDAPException as e:
            logger.error(f"Failed to connect to AD: {e}")
            return False

    def disconnect(self):
        """Close AD connection"""
        if self.conn:
            self.conn.unbind()
            self.conn = None
            logger.info("Disconnected from AD")

    def get_all_users(self, offset: int = 0, limit: int = 100, include_photo: bool = True) -> tuple[list[dict], int]:
        """Fetch users from AD with pagination"""
        if not self.conn:
            if not self.connect():
                return [], 0

        try:
            search_filter = "(&(objectClass=user)(objectCategory=person)(!(userAccountControl:1.2.840.113556.1.4.803:=2)))"

            self.conn.search(
                search_base=self.config['users_ou'],
                search_filter=search_filter,
                search_scope=SUBTREE,
                attributes=self.config['attributes']
            )

            all_entries = self.conn.entries
            total = len(all_entries)

            # Apply pagination
            paginated = all_entries[offset:offset + limit]

            users = []
            for entry in paginated:
                user = self._parse_user(entry, include_photo=include_photo)
                if user:
                    users.append(user)

            logger.info(f"Fetched {len(users)} users from AD (offset={offset}, limit={limit}, total={total})")
            return users, total

        except LDAPException as e:
            logger.error(f"Failed to fetch users: {e}")
            return [], 0

    def get_user_by_email(self, email: str) -> Optional[dict]:
        """Fetch single user by email"""
        if not self.conn:
            if not self.connect():
                return None

        try:
            search_filter = f"(&(objectClass=user)(mail={email}))"

            self.conn.search(
                search_base=self.config['base_dn'],
                search_filter=search_filter,
                search_scope=SUBTREE,
                attributes=self.config['attributes']
            )

            if self.conn.entries:
                return self._parse_user(self.conn.entries[0])
            return None

        except LDAPException as e:
            logger.error(f"Failed to fetch user {email}: {e}")
            return None

    def get_user_by_dn(self, dn: str) -> Optional[dict]:
        """Fetch user by Distinguished Name"""
        if not self.conn:
            if not self.connect():
                return None

        try:
            self.conn.search(
                search_base=dn,
                search_filter="(objectClass=user)",
                search_scope=SUBTREE,
                attributes=self.config['attributes']
            )

            if self.conn.entries:
                return self._parse_user(self.conn.entries[0])
            return None

        except LDAPException as e:
            logger.error(f"Failed to fetch user by DN {dn}: {e}")
            return None

    def get_subordinates(self, manager_dn: str) -> list[dict]:
        """Get direct reports for a manager"""
        if not self.conn:
            if not self.connect():
                return []

        try:
            search_filter = f"(&(objectClass=user)(manager={manager_dn}))"

            self.conn.search(
                search_base=self.config['base_dn'],
                search_filter=search_filter,
                search_scope=SUBTREE,
                attributes=self.config['attributes']
            )

            return [self._parse_user(entry) for entry in self.conn.entries if self._parse_user(entry)]

        except LDAPException as e:
            logger.error(f"Failed to fetch subordinates: {e}")
            return []

    def authenticate_user(self, username: str, password: str) -> Optional[dict]:
        """Authenticate user against AD"""
        try:
            server = Server(
                self.config['server'],
                port=self.config['port'],
                use_ssl=self.config['use_ssl']
            )

            # Try to bind with user credentials
            user_conn = Connection(
                server,
                user=username,
                password=password,
                authentication=NTLM,
                auto_bind=True
            )

            # If successful, get user info
            # Extract username without domain (can't use backslash in f-string)
            sam_account = username.split('\\')[-1] if '\\' in username else username
            user_conn.search(
                search_base=self.config['base_dn'],
                search_filter=f"(sAMAccountName={sam_account})",
                attributes=self.config['attributes']
            )

            if user_conn.entries:
                user = self._parse_user(user_conn.entries[0])
                user_conn.unbind()
                logger.info(f"User authenticated: {username}")
                return user

            user_conn.unbind()
            return None

        except LDAPException as e:
            logger.warning(f"Authentication failed for {username}: {e}")
            return None

    def _parse_user(self, entry, include_photo: bool = False) -> Optional[dict]:
        """Parse LDAP entry to user dict"""
        try:
            # Get photo as base64 only if requested (photos are large!)
            photo = None
            if include_photo and hasattr(entry, 'thumbnailPhoto') and entry.thumbnailPhoto.value:
                photo = base64.b64encode(entry.thumbnailPhoto.value).decode('utf-8')

            # Extract manager email from DN
            manager_dn = str(entry.manager) if hasattr(entry, 'manager') and entry.manager else None

            return {
                'dn': str(entry.entry_dn),
                'name': str(entry.cn) if hasattr(entry, 'cn') else None,
                'email': str(entry.mail) if hasattr(entry, 'mail') else None,
                'login': str(entry.sAMAccountName) if hasattr(entry, 'sAMAccountName') else None,
                'upn': str(entry.userPrincipalName) if hasattr(entry, 'userPrincipalName') else None,
                'title': str(entry.title) if hasattr(entry, 'title') else None,
                'department': str(entry.department) if hasattr(entry, 'department') else None,
                'manager_dn': manager_dn,
                'photo_base64': photo
            }
        except Exception as e:
            logger.error(f"Failed to parse user entry: {e}")
            return None

    def build_org_tree(self, users: list[dict]) -> dict:
        """Build organization tree from flat user list"""
        # Create lookup by DN
        by_dn = {u['dn']: u for u in users}

        # Add children lists
        for user in users:
            user['subordinates'] = []

        # Build tree
        roots = []
        for user in users:
            if user['manager_dn'] and user['manager_dn'] in by_dn:
                by_dn[user['manager_dn']]['subordinates'].append(user)
            else:
                roots.append(user)

        return {
            'roots': roots,
            'total_count': len(users)
        }


# Test connection
if __name__ == "__main__":
    import yaml
    import os

    logging.basicConfig(level=logging.DEBUG)

    with open('config.yaml', 'r') as f:
        config = yaml.safe_load(f)

    # Expand env vars
    config['ad']['bind_user'] = os.environ.get('AD_BIND_USER', config['ad']['bind_user'])
    config['ad']['bind_password'] = os.environ.get('AD_BIND_PASSWORD', config['ad']['bind_password'])

    client = ADClient(config)
    if client.connect():
        users = client.get_all_users()
        print(f"Found {len(users)} users")
        for u in users[:5]:
            print(f"  - {u['name']} ({u['email']})")
        client.disconnect()

File: connector-app/src-tauri/src/ad_client.rs
use ldap3::{Ldap, LdapConnAsync, LdapConnSettings, Scope, SearchEntry, controls::{MakeCritical, SimplePagedResults}};
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use tracing::{error, info};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ADUser {
    pub dn: String,
    pub name: Option<String>,
    pub email: Option<String>,
    pub login: Option<String>,
    pub title: Option<String>,
    pub department: Option<String>,
    pub manager_dn: Option<String>,
    pub phone: Option<String>,
    pub mobile: Option<String>,
    pub photo_base64: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct SyncStats {
    pub total_in_ad: usize,
    pub with_department: usize,
    pub without_department: usize,
    pub with_email: usize,
    pub without_email: usize,
    pub filtered_out: usize,
    pub returned: usize,
}

pub struct ADClient {
    server: String,
    port: u16,
    use_ssl: bool,
    bind_user: String,
    bind_password: String,
    base_dn: String,
    users_ou: String,
}

impl ADClient {
    pub fn new(
        server: &str,
        port: u16,
        use_ssl: bool,
        bind_user: &str,
        bind_password: &str,
        base_dn: &str,
    ) -> Self {
        // Derive users OU from base_dn
        let users_ou = format!("OU=EKF-USERS,{}", base_dn);

        Self {
            server: server.to_string(),
            port,
            use_ssl,
            bind_user: bind_user.to_string(),
            bind_password: bind_password.to_string(),
            base_dn: base_dn.to_string(),
            users_ou,
        }
    }

    async fn connect(&self) -> Result<Ldap, String> {
        let url = if self.use_ssl {
            format!("ldaps://{}:{}", self.server, self.port)
        } else {
            format!("ldap://{}:{}", self.server, self.port)
        };

        info!("Connecting to AD: {}", url);

        let settings = LdapConnSettings::new();
        let (conn, mut ldap) = LdapConnAsync::with_settings(settings, &url)
            .await
            .map_err(|e| format!("Failed to connect to AD: {}", e))?;

        // Spawn connection handler
        tokio::spawn(async move {
            if let Err(e) = conn.drive().await {
                error!("LDAP connection error: {}", e);
            }
        });

        // Bind with credentials
        ldap.simple_bind(&self.bind_user, &self.bind_password)
            .await
            .map_err(|e| format!("Failed to bind to AD: {}", e))?
            .success()
            .map_err(|e| format!("Bind failed: {}", e))?;

        info!("Connected to AD successfully");
        Ok(ldap)
    }

    pub async fn test_connection(&self) -> Result<(), String> {
        let mut ldap = self.connect().await?;
        ldap.unbind().await.ok();
        Ok(())
    }

    pub async fn get_all_users(
        &self,
        require_department: bool,
        require_email: bool,
        include_photo: bool,
    ) -> Result<(Vec<ADUser>, SyncStats), String> {
        let mut ldap = self.connect().await?;

        // Search for active users
        let filter = "(&(objectClass=user)(objectCategory=person)(!(userAccountControl:1.2.840.113556.1.4.803:=2)))";

        let mut attrs = vec![
            "cn",
            "mail",
            "sAMAccountName",
            "userPrincipalName",
            "title",
            "department",
            "manager",
            "telephoneNumber",
            "mobile",
        ];

        if include_photo {
            attrs.push("thumbnailPhoto");
        }

        let mut stats = SyncStats {
            total_in_ad: 0,
            with_department: 0,
            without_department: 0,
            with_email: 0,
            without_email: 0,
            filtered_out: 0,
            returned: 0,
        };

        let mut users = Vec::new();

        // Use paged search to handle large directories (page size 500)
        let mut page_ctrl = SimplePagedResults::new(500, false);

        loop {
            let controls = vec![page_ctrl.clone().critical().into()];

            let search_result = ldap
                .with_controls(controls)
                .search(&self.base_dn, Scope::Subtree, filter, attrs.clone())
                .await
                .map_err(|e| format!("Search failed: {}", e))?;

            let (rs, res) = search_result
                .success()
                .map_err(|e| format!("Search result error: {}", e))?;

            stats.total_in_ad += rs.len();

            for entry in rs {
                let entry = SearchEntry::construct(entry);

                let department = entry.attrs.get("department").and_then(|v| v.first().cloned());
                let email = entry.attrs.get("mail").and_then(|v| v.first().cloned());

                // Update stats
                if department.is_some() {
                    stats.with_department += 1;
                } else {
                    stats.without_department += 1;
                }

                if email.is_some() {
                    stats.with_email += 1;
                } else {
                    stats.without_email += 1;
                }

                // Apply filters
                if require_department && department.is_none() {
                    stats.filtered_out += 1;
                    continue;
                }

                if require_email && email.is_none() {
                    stats.filtered_out += 1;
                    continue;
                }

                // Parse photo
                let photo_base64 = if include_photo {
                    entry
                        .bin_attrs
                        .get("thumbnailPhoto")
                        .and_then(|photos| photos.first())
                        .map(|photo| base64::Engine::encode(&base64::engine::general_purpose::STANDARD, photo))
                } else {
                    None
                };

                let user = ADUser {
                    dn: entry.dn,
                    name: entry.attrs.get("cn").and_then(|v| v.first().cloned()),
                    email,
                    login: entry.attrs.get("sAMAccountName").and_then(|v| v.first().cloned()),
                    title: entry.attrs.get("title").and_then(|v| v.first().cloned()),
                    department,
                    manager_dn: entry.attrs.get("manager").and_then(|v| v.first().cloned()),
                    phone: entry.attrs.get("telephoneNumber").and_then(|v| v.first().cloned()),
                    mobile: entry.attrs.get("mobile").and_then(|v| v.first().cloned()),
                    photo_base64,
                };

                users.push(user);
            }

            // Check if there are more pages
            if let Some(cookie) = SimplePagedResults::from(&res) {
                if cookie.is_empty() {
                    break;
                }
                page_ctrl.set_cookie(cookie);
            } else {
                break;
            }
        }

        stats.returned = users.len();

        ldap.unbind().await.ok();

        info!(
            "AD sync: {} total, {} with dept, {} filtered, {} returned",
            stats.total_in_ad, stats.with_department, stats.filtered_out, stats.returned
        );

        Ok((users, stats))
    }

    pub async fn authenticate(&self, username: &str, password: &str) -> Result<Option<ADUser>, String> {
        // Try to bind with user credentials
        let url = if self.use_ssl {
            format!("ldaps://{}:{}", self.server, self.port)
        } else {
            format!("ldap://{}:{}", self.server, self.port)
        };

        let settings = LdapConnSettings::new();
        let (conn, mut ldap) = LdapConnAsync::with_settings(settings, &url)
            .await
            .map_err(|e| format!("Failed to connect: {}", e))?;

        tokio::spawn(async move {
            conn.drive().await.ok();
        });

        // Try NTLM-style bind
        let result = ldap.simple_bind(username, password).await;

        match result {
            Ok(res) => {
                if res.rc == 0 {
                    // Success - now fetch user info
                    let sam = username.split('\\').last().unwrap_or(username);
                    let filter = format!("(sAMAccountName={})", sam);

                    let (rs, _) = ldap
                        .search(
                            &self.base_dn,
                            Scope::Subtree,
                            &filter,
                            vec!["cn", "mail", "title", "department", "manager"],
                        )
                        .await
                        .map_err(|e| format!("Search failed: {}", e))?
                        .success()
                        .map_err(|e| format!("Search error: {}", e))?;

                    ldap.unbind().await.ok();

                    if let Some(entry) = rs.into_iter().next() {
                        let entry = SearchEntry::construct(entry);
                        return Ok(Some(ADUser {
                            dn: entry.dn,
                            name: entry.attrs.get("cn").and_then(|v| v.first().cloned()),
                            email: entry.attrs.get("mail").and_then(|v| v.first().cloned()),
                            login: Some(sam.to_string()),
                            title: entry.attrs.get("title").and_then(|v| v.first().cloned()),
                            department: entry.attrs.get("department").and_then(|v| v.first().cloned()),
                            manager_dn: entry.attrs.get("manager").and_then(|v| v.first().cloned()),
                            phone: None,
                            mobile: None,
                            photo_base64: None,
                        }));
                    }
                }
                ldap.unbind().await.ok();
                Ok(None)
            }
            Err(_) => {
                ldap.unbind().await.ok();
                Ok(None)
            }
        }
    }
}
